File: app/src/main/java/com/example/androidcarapp/api/CarEncoder.kt
package com.example.androidcarapp.api

/**
 * 小车控制指令编码器 (1:1 对应 HarmonyOS CarEncode)
 *
 * 协议格式: $01<type><size><data><checksum>#
 * 所有数值以十六进制大写字符串表示
 */
object CarEncoder {

    private const val TAG = "CarEncoder"

    /** 按钮控制编码 */
    fun buttonEncode(d: CarDirection): String {
        return baseEncode("15", numberToHex(d.code, 2))
    }

    /** 摇杆控制编码 (speedX, speedY: -100~100) */
    fun ctrlEncode(speedX: Float, speedY: Float): String {
        var x = Math.round(speedX)
        var y = Math.round(speedY)

        if (x < 0) x += 256
        if (y < 0) y += 256

        return baseEncode("10", numberToHex(x, 2) + numberToHex(y, 2))
    }

    /** 启动跟踪 */
    fun trackingOpen(): String = baseEncode("63")

    /** 关闭跟踪 */
    fun trackingClose(): String = baseEncode("64")

    /** 通用编码函数 */
    private fun baseEncode(type: String, vararg datas: String): String {
        val info = datas.joinToString("")
        val size = numberToHex(info.length + 2, 2)
        var code = "01$type$size$info"
        code += numberToHex(checksum(code), 2)
        return "$$code#"
    }

    /** 数值转十六进制字符串 */
    private fun numberToHex(num: Int, len: Int): String {
        val hex = num.toString(16)
        return hex.padStart(len, '0').uppercase()
    }

    /** 计算校验和 */
    private fun checksum(data: String): Int {
        var sum = 0
        var i = 0
        while (i < data.length) {
            sum += data.substring(i, i + 2).toInt(16)
            i += 2
        }
        return sum % 256
    }
}

File: web-admin/server/carEncoder.js
/**
 * 小车 TCP 协议编码（与 Android CarEncoder.kt 帧格式一致）
 * 帧格式: $01<type><size><data><checksum>#
 */

function numberToHex(num, len) {
  return num.toString(16).padStart(len, '0').toUpperCase();
}

function checksum(data) {
  let sum = 0;
  for (let i = 0; i < data.length; i += 2) {
    sum += parseInt(data.substring(i, i + 2), 16);
  }
  return sum % 256;
}

function baseEncode(type, ...datas) {
  const info = datas.join('');
  const size = numberToHex(info.length + 2, 4);
  let code = `01${type}${size}${info}`;
  code += numberToHex(checksum(code), 2);
  return `$${code}#`;
}

function stringToHex(str) {
  return Buffer.from(str, 'utf8').toString('hex').toUpperCase();
}

/** type=20 订单下发，data 为 UTF-8 JSON 的十六进制编码 */
function encodeOrderFrame(orderPayload) {
  const json = JSON.stringify(orderPayload);
  return baseEncode('20', stringToHex(json));
}

/** type=21 订单取消 */
function encodeCancelFrame(orderNo) {
  return baseEncode('21', stringToHex(JSON.stringify({ action: 'cancel', order_no: orderNo })));
}

/** type=22 人脸扫描指令 */
function encodeFaceScanFrame(scanPayload) {
  const json = JSON.stringify(scanPayload);
  return baseEncode('22', stringToHex(json));
}

module.exports = { encodeOrderFrame, encodeCancelFrame, encodeFaceScanFrame, baseEncode };
